### ISSUE TITLE
How to test this package in 6.X+?

### ISSUE THREAD

**Issue by carusology** [auth0/auth0.net#250]
### Description

We've been using the `5.X` versions of this package for some time as part of our user management workflows in our ASP.NET Core v2.1 product. We have built unit tests over those libraries by mocking the various `I*Client` interfaces such as `IConnectionsClient`, `IUsersClient`, etc. Now I see the interfaces are retired, and I don't understand why. 

The [nuget page](https://www.nuget.org/packages/Auth0.ManagementApi/) states that this is so "All I*Client interfaces have been removed so adding endpoints is no longer breaking." There is a link to a migration guide (a dead link, but I assume is supposed to point [here](http://auth0.github.io/auth0.net/migrating.html#migrating-from-version-5-to-version-6)?), which states a surprising comment: " As this library is a concrete implementation of calling the Auth0 management APIs and not a general abstraction layer over management this interface served no purpose." I have to respectfully disagree here. Us, the consumers of your API, use and wrap these interfaces with composition to alter behavior, set default values from configuration, manage rate-limiting, add logging, etc. Remove the interface, and you remove our ability to compose these base implementations.

The end result of that decision is now I'm looking at this very major change and I'm left scratching my head with a straightforward question: **How are we intended to write tests for this code?**

If we are to avoid unit tests in favor of integration tests, Auth0 is rate-limited, so the tests will take forever to run and we have to keep track of how many requests are going out a second. This kind of tracking as a side-effect is pretty easily done with composition, but composition in .NET is traditionally done via interface. We also can't use tools such as Moq to mock the interface because, again, no interface exists. **Are we intended to use partial mocks? Is that how it was envisioned that we would write tests?**

For what we should do on our end, I'm left thinking we should locally reimplement the interfaces that were just removed and use these various `*Clients` classes as the base implementations of those interfaces. But if we're doing that, it goes back to my original confusion as to why these interfaces were removed in the first place, and how the maintainers expect us to test the various clients in our own code bases. Maybe this _is_ the intention? Should we write our own interfaces around the functionally we use?

Anyway, I'd love some insight into how consumers should be thinking about their testing strategies for this library as a result of this change.

### Prerequisites

[X] I have checked the [documentation for this library](https://auth0.github.io/auth0.net/).

[X] I have checked the [Auth0 Community](https://community.auth0.com/) for related posts.

[X] I have checked for related or duplicate [Issues](https://github.com/auth0/auth0.net/issues) and [PRs](https://github.com/auth0/auth0.net/pulls).

[X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md).

~~[ ] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md).~~ This is a dead link.

### Environment

Please provide the following:
* .NET Core v2.1 / v2.2
* Auth0.NET 5.X / 6.0

**Comment by damieng** [comment]
Sorry the link to the migration guide was accidentally pointing to the release PR branch, it should have been pointing at master http://auth0.github.io/auth0.net/migrating.html

The problem with having these interfaces is that we can not add a single endpoint without it being a breaking change.  e.g.

- We add permissions... that's three more methods on IRolesClient.  Which is a breaking change.
- New GuardianClient?  Requires a new property added to IMangementApi - breaking change.

Indeed, if the interfaces were still in 6.0 you project would still not compile because of these two changes. Even though you don't necessarily care about them.

This is not a great position to be in. It is disruptive for customers who just want the latest bits and for us to continually have to bump the major version to comply with semantic versioning expectations and write a short migration guide every release.

That said I appreciate that many people need to mock remote calls for testing and we can provide a bit more guidance around that.  Right now people who have used these interfaces for this purpose must already have their own in-memory alternatives.

Ideally people will replace IManagementClient with an interface that exposes just the functionality they want to use.  One implementation will go to in-memory the other will pass through to ManagementClient.

This way as we add new features nobody gets broken builds.

**Comment by carusology** [comment]
Thanks for the prompt and empathetic response, @damieng.

> Indeed, if the interfaces were still in 6.0 your project would still not compile because of these two changes. Even though you don't necessarily care about them.

I have to disagree here and say this is a strawman.

If I don't implement these interfaces, my code would actually continue to compile. To use the example you gave, I don't care about guardian or role particular changes, and since I don't implement those interfaces, my code would still have compiled as the interfaces I did not care about are the only ones that changed. So, unfortunately, the 6.0 breaking change is actually causing the inverse of the described goal. I don't care about those changes, yet my build is breaking, and in a very big way. I have a lot more work to do (recreate about ~5 interfaces and ~5 base implementations) that I would not have to do otherwise (update the version number in my *.csproj file).

In truth, my build would only break if the breaking change was to an Auth0 interface that I have implemented changes. But isn't that exactly when I want it to break? Auth0 changed the contract of an endpoint I utilize, and it's on me to update it. I don't own that interface. If I implement the roles interface, I presumably _do_ care about them, because I implemented it.

The way to limit the impact of these types of changes is to keep interfaces small and focused on the smallest useful areas of responsibility (which Auth0 actually did _really well_!). It isn't to remove the interfaces entirely.

> Ideally people will replace IManagementClient with an interface that exposes just the functionality they want to use. One implementation will go to in-memory the other will pass through to ManagementClient.

If a customer cared about preventing interface updates from breaking builds, there was nothing stopping them from doing this in 5.X+. We could wrap your interfaces in the manner you described here. However, by removing them from the library, you're forcing us to do this, instead of letting it be a conscious choice on our side. Now all of us have to wrap, even those of us who to defer to Auth0 to define these interfaces for us.

So now for every interface I use, I have to rewrite each client's interface and inject the base implementation into a base implementation of that interface. That's a lot more code I didn't have to write before.

I hope you reconsider the removal of these interfaces, and include them once again in this repository.

**Comment by damieng** [comment]
> I don't care about guardian or role particular changes, and since I don't implement those interfaces, my code would still have compiled as the interfaces.

There are three scenarios:

1. Did not use any interfaces - this change breaks nothing
2. Use IManagementApi - this release would have broken you whether we removed the interfaces or not
3. Used IRolesClient - this release would have broken you regardless
4. Used select I*Client interfaces except IRolesClient

The only scenario where interface removal caused pain that wouldn't have broken otherwise is if you were using some of the I*Client interfaces but not IManagementApi or IRolesClient.  I'm assuming that is the scenario you are in?

> If a customer cared about preventing interface updates from breaking builds,

There is a clear expectation in our industry that patch and minor updates will not break existing code. The previous interface structures did not allow that.  Any new endpoint was a break.

> small and focused on the smallest useful areas of responsibility

I agree with that statement, this would be a way we could re-introduce interfaces.

IManagementApi would never exist again but smaller focused I*Client replacements could exist for sets of related operations at a more granular level.  e.g. IRolesClient would not exist as it did before because it attempts to capture everything in the /roles endpoint which today is three sets of related things (Roles, Permissions for Roles and Users for Roles).  We would instead have three small interfaces to replace them.

Some other endpoints that would likely need more fine-grained interfaces would include Jobs, Stats, Users, Guardian, Tickets.

**Comment by BrianMikinski** [comment]
@damieng 

I'm encountering the same issue as @carusology in regards to unit testing. In scenarios 1-3, I don't really have any problem with breaking changes when I bump my major version - in fact I expect that to happen in those scenarios. I regards to the 4th example, I still don't have any issues with breaking changes but having an API that matches the IClient interfaces greases the wheels when unit testing. It also reduces the amount of code a developer has to write in order to unit test the their project and ensures that the  developer create interface is not an incomplete subset of the Auth0 offering. I think devs would appreciate having interfaces but understand if the decision is to get rid of them.

**Comment by damieng** [comment]
As I have said before I'm up for introducing interfaces on a more granular level.  i.e. having a IUsersClient is a no-go as it has 18 methods however we could have an IUserRolesClient, IUserLinkingClient, IUserPermissionsClient, IUserReadClient and IUserWriteClient for example.

This way at least as we introduce new blocks of functionality nothing would break as they would likely come under a new interface - much as IUserRolesClient and IUserPermissionsClient would have recently.

For backward compatibility and because our endpoints are grouped into large blocks on the REST/Swagger side UsersClient would still exist and implement these interfaces.

**Comment by esbenbach** [comment]
I am also encountering this, and for my part I would actually have been happy if the UsersClient and ManagementApiClient would take in IApiConnection in the ctor, and make all the public methods virtual.

I would prefer interfaces, but lacking that if I can stub/mock the implementations that is "good enough", right now, I can't so I have to do as @carusology and introduce interfaces and wrapper classes for my testing in otherwise known working code.

I can live with the fact that my DI container and dependencies knows the concrete types. But the situation as it is, is just horrible.

**Comment by dandenton** [comment]
@damieng I'm not sure I understand the reasoning for not wanting an IManagementApi and the comment about a new endpoint being a breaking change.  As long as an existing endpoint isn't removed or modified how does adding a new method on the interface constitute a breaking change?  Even adding a new I*Client with an associated property on IManagementApi wouldn't be a breaking change since no one would have been referencing it unless it was a refactor of an existing client.

**Comment by damieng** [comment]
It is impossible for us to add any new endpoint with the old `IXClient` and `IManagementApi` interfaces without breaking changes that stop your app compiling with the new version. Either:

1. We add an endpoint to an existing `IxxxClient` interface and you had already implemented that interface.  Now your app won't compile because you don't implement that new interface method we added.

2. We add a new set of endpoints with a new `IxxxClient` which we then have to add to `IManagementApi` and you implemented `IManagementApi` but now don't expose that new Client so again fail to compile.

In addition to that the IxxxClient interfaces expose concrete dto/models etc. and it becomes quite the sprawl.

I think the simplest and most flexible solution is we modify ManagementClient to have a new constructor that takes an `IApiConnection`.  Then you could just create your own mock/stub that implements IApiConnection and does whatever you need it to do for the classes you care about (or indeed just retain posts in memory and return them on gets etc).

Here's https://github.com/auth0/auth0.net/blob/master/src/Auth0.Core/Http/IApiConnection.cs which has just the 7 methods you'd need to be able to mock anything and we'd never break people with new endpoints.

**Comment by robertmclaws** [comment]
One option might be to use the new Default Interface implementation in C# 8.0 to add default implementations when new interfaces are added. AFAIK that doesn't break existing users because there is code to backstop the interface.

**Comment by damieng** [comment]
Hey Robert.

I did consider default implementations but there isn't any kind of default that would make sense on the IxxxClient interfaces - each one is literally a completely different operation.

I think the IApiConnection abstraction works particularly well for mocking as you can just create an IApiConnection mock/stub that returns whatever result your unit test was hoping for.

I could actually make this even simpler if I evolve IApiConnection so that instead of having multiple methods - one for each http operation type - it simply has just one extra parameter defining the method type, e.g.

```csharp
Task<T> SendRequestAsync<T>(HttpMethod method, string resourceUri, object body = null, IDictionary<string, object> query = null);
```

**Comment by yannlairdc** [comment]
You do realise that has of now, If I want to mock something like `managementClient.Users.GetUsersByEmailAsync(xxx)` it is impossible? Unless I missed something of course, but I have look at the implementation of the ManagementApiClient and can't find a solution.

No interfaces or parameter-less constructors  => No mocking possible -> Can't unit test code using the managementApiClient. 
This is kinda a not going in the same direction as the .Net core world and the huge push toward DI and micro services. 

If I want to test it:
-  I actually need to create integration tests with a test instance of Auth0 that I can initiate with an actual Token and Management domain -> defeat the purpose of unit tests

- I have to write my own wrapper around your client with a 1-to-1 match to each single action that I need to use -> basically forcing every single customers who want to unit tests to re implement interfaces themselves.

Could you please provide everyone reading this topic with a simple and cost effective way of doing unit tests? (If possible with a written example)

**Comment by damieng** [comment]
Merged in, will ship in 7.0.0 - docs on how to use it will be in the migration guide.
